src/lib/y2packager/release_notes_reader.rb

```diff
@@ -11,170 +11,133 @@
+  #   This behaviour covers the case in which you are installing behind a SMT but without
+  #   access to Internet.
+  # * When the system *is not registered*: it will work the other way around, trying
+  #   first relnotes_url and falling back to RPM packages.
+  #
+  # ### Cached release notes
+  #
+  # Release notes are stored using an instance of `Y2Packager::ReleaseNotesStore`.
+  # When trying to read a product license for second time, this class will try to fetch
+  # the latest version (determined by ReleaseNotesRpmReader#latest_version or
+  # ReleaseNotesUrlReader#latest_version from the store). If release notes are not there,
+  # or the stored version is outdated (maybe a new package is now available), it will
+  # try to get that version.
+  #
+  # Take into account that, when using the relnotes_url property, a URL that already
+  # failed will not be retried again. See ReleaseNotesUrlReader for further details.
   class ReleaseNotesReader
     include Yast::Logger
 
+    # Product to get release notes for
+    attr_reader :product
+
     # Constructor
     #
+    # @param product             [Product]           Product to get release notes for
     # @param release_notes_store [ReleaseNotesStore] Release notes store to cache data
-    def initialize(release_notes_store = nil)
+    def initialize(product, release_notes_store = nil)
       @release_notes_store = release_notes_store
+      @product = product
     end
 
+    # Fallback language
+    FALLBACK_LANG = "en".freeze
+
     # Get release notes for a given product
     #
-    # Release notes are downloaded and extracted to work directory.  When
-    # release notes for a language "xx_XX" are not found, it will fallback to
-    # "xx".
-    #
-    # @param product   [Y2Packager::Product] Product
-    # @param user_lang [String]              Release notes language (falling back to "en_US"
-    #                                        and "en")
-    # @param format    [Symbol]              Release notes format (:txt or :rtf)
-    # @return [String,nil] Release notes or nil if a release notes were not found
+    # @param user_lang [String] Release notes language (falling back to "en")
+    # @param format    [Symbol] Release notes format (:txt or :rtf)
+    # @return [String,nil] Release notes or nil if release notes were not found
     #   (no package providing release notes or notes not found in the package)
-    def release_notes_for(product, user_lang: "en_US", format: :txt)
-      package = release_notes_package_for(product)
-      if package.nil?
-        log.info "No package containing release notes for #{product.name} was found"
-        return nil
+    def release_notes(user_lang: "en_US", format: :txt)
+      readers =
+        # registered system: get relnotes from RPMs and fallback to relnotes_url property
+        if registered?
+          [
+            ReleaseNotesRpmReader.new(product),
+            ReleaseNotesUrlReader.new(product)
```

Comment:
Well, this suggests that these 2 classes should have a superclass that documents their interface
(and the #latest_version type of [String,:latest] would clarify my earlier question).

But an obvious name ReleaseNotesReader is already taken for something else.

Reply:
Yes, we could use `ReleaseNotesFetcher` (with `ReleaseNotesRpmFetcher` and `ReleaseNotesUrlFetcher`) or `ReleaseNotesStrategy` or something like that. I will try to come up with some name which makes sense. I am also open to suggestions :)